extendr-api/src/robj/try_from_robj.rs

```diff
@@ -2,54 +2,95 @@
 
 use super::*;
 
-macro_rules! impl_try_from_scalar {
-    ($t: ty) => {
+macro_rules! impl_try_from_scalar_integer {
+    ($t:ty) => {
         impl TryFrom<Robj> for $t {
             type Error = Error;
 
             fn try_from(robj: Robj) -> Result<Self> {
-                if let Some(v) = robj.as_integer_slice() {
-                    match v.len() {
-                        0 => Err(Error::ExpectedNonZeroLength(robj)),
-                        1 => {
-                            if !v[0].is_na() {
-                                Ok(v[0] as Self)
-                            } else {
-                                Err(Error::MustNotBeNA(robj))
-                            }
-                        }
-                        _ => Err(Error::ExpectedScalar(robj)),
+                // Check if the value is a scalar
+                match robj.len() {
+                    0 => return Err(Error::ExpectedNonZeroLength(robj)),
+                    1 => {}
+                    _ => return Err(Error::ExpectedScalar(robj)),
+                };
+
+                // Check if the value is not a missing value
+                if robj.is_na() {
+                    return Err(Error::MustNotBeNA(robj));
+                }
+
+                // If the conversion is int-to-int, check the limits. This
+                // needs to be down by `TryFrom` because the conversion by `as`
+                // is problematic on unsigned integer types.
+                if let Some(v) = robj.as_integer() {
+                    if let Ok(v) = Self::try_from(v) {
+                        return Ok(v);
+                    } else {
+                        return Err(Error::OutOfLimits(robj));
                     }
-                } else if let Some(v) = robj.as_real_slice() {
-                    match v.len() {
-                        0 => Err(Error::ExpectedNonZeroLength(robj)),
-                        1 => {
-                            if !v[0].is_na() {
-                                Ok(v[0] as Self)
-                            } else {
-                                Err(Error::MustNotBeNA(robj))
-                            }
-                        }
-                        _ => Err(Error::ExpectedScalar(robj)),
+                }
+
+                // If the conversion is float-to-int, check if the value is
+                // integer-ish. This needs to be down with `as`, as no `TryFrom`
+                // is implemented for float types. `FloatToInt` trait might
+                // eventually become available in future, though.
+                if let Some(v) = robj.as_real() {
+                    let result = v as Self;
+                    if ((result as f64 - v).abs() < f64::EPSILON) {
```

Comment:
Didn't @andy-thomason suggest converting from float to integer back to float and testing for equality with the original value? Not sure which way is better.

Reply:
Yes, I first implemented so, but clippy warned.

https://rust-lang.github.io/rust-clippy/master/#float_cmp

Reply:
Funny, because in this case we do actually want to test for exact equality. Anyways, your current implementation looks fine to me, so let's just leave it as is.

Reply:
I think what the current code does is just the verbose version of checking "exact equality." When the difference is unrepresentable even by the machine epsilon, two values should be the same. So, I think we can choose to silence clippy here, but this code is probably a good reminder about the trickiness of comparing two floating point numbers.